File: godtools-sync/src/main/java/org/cru/godtools/sync/LanguagesSyncTasks.java
package org.cru.godtools.sync;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.util.SimpleArrayMap;

import com.annimon.stream.Collectors;

import org.ccci.gto.android.common.db.Query;
import org.ccci.gto.android.common.jsonapi.model.JsonApiObject;
import org.ccci.gto.android.common.jsonapi.retrofit2.JsonApiParams;
import org.cru.godtools.model.Language;

import java.io.IOException;
import java.util.Locale;
import java.util.Map;

import retrofit2.Response;

import static org.ccci.gto.android.common.TimeConstants.WEEK_IN_MS;

final class LanguagesSyncTasks extends BaseDataSyncTasks {
    private static final Object LOCK_SYNC_LANGUAGES = new Object();

    private static final String SYNC_TIME_LANGUAGES = "last_synced.languages";
    private static final long STALE_DURATION_LANGUAGES = WEEK_IN_MS;

    LanguagesSyncTasks(@NonNull final Context context) {
        super(context);
    }

    boolean syncLanguages(@NonNull final Bundle args) throws IOException {
        final SimpleArrayMap<Class<?>, Object> events = new SimpleArrayMap<>();

        synchronized (LOCK_SYNC_LANGUAGES) {
            // short-circuit if we aren't forcing a sync and the data isn't stale
            final boolean force = isForced(args);
            if (!force &&
                    System.currentTimeMillis() - mDao.getLastSyncTime(SYNC_TIME_LANGUAGES) < STALE_DURATION_LANGUAGES) {
                return true;
            }

            // fetch languages from the API
            // short-circuit if this response is invalid
            final Response<JsonApiObject<Language>> response = mApi.languages.list(new JsonApiParams()).execute();
            if (response == null || response.code() != 200) {
                return false;
            }

            // store languages
            final JsonApiObject<Language> json = response.body();
            if (json != null) {
                final Map<Locale, Language> existing = mDao.streamCompat(Query.select(Language.class))
                        .collect(Collectors.toMap(Language::getCode, l -> l));
                storeLanguages(events, json.getData(), existing);
            }

            // send any pending events
            sendEvents(events);

            // update the sync time
            mDao.updateLastSyncTime(SYNC_TIME_LANGUAGES);
        }

        return true;
    }
}

File: library/sync/src/main/java/org/cru/godtools/sync/task/LanguagesSyncTasks.kt
package org.cru.godtools.sync.task

import android.os.Bundle
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import org.ccci.gto.android.common.base.TimeConstants
import org.ccci.gto.android.common.db.Expression.Companion.constants
import org.ccci.gto.android.common.db.Query
import org.ccci.gto.android.common.jsonapi.retrofit2.JsonApiParams
import org.cru.godtools.api.LanguagesApi
import org.cru.godtools.model.Language
import org.greenrobot.eventbus.EventBus
import org.keynote.godtools.android.db.Contract.LanguageTable
import org.keynote.godtools.android.db.GodToolsDao
import timber.log.Timber

private const val SYNC_TIME_LANGUAGES = "last_synced.languages"
private const val STALE_DURATION_LANGUAGES = TimeConstants.WEEK_IN_MS

@Singleton
class LanguagesSyncTasks @Inject internal constructor(
    dao: GodToolsDao,
    private val languagesApi: LanguagesApi,
    eventBus: EventBus
) : BaseDataSyncTasks(dao, eventBus) {
    private val languagesMutex = Mutex()

    suspend fun syncLanguages(args: Bundle) = withContext(Dispatchers.IO) {
        languagesMutex.withLock {
            // short-circuit if we aren't forcing a sync and the data isn't stale
            if (!isForced(args) &&
                System.currentTimeMillis() - dao.getLastSyncTime(SYNC_TIME_LANGUAGES) < STALE_DURATION_LANGUAGES
            ) return@withContext true

            // fetch & store languages
            languagesApi.list(JsonApiParams()).takeIf { it.isSuccessful }?.body()?.let { json ->
                dao.transaction {
                    val existing = dao.get(Query.select<Language>())
                        .groupingBy { it.code }
                        .reduce { _, lang1, lang2 ->
                            Timber.tag("LanguagesSyncTask").d(
                                RuntimeException("Duplicate Language sync error"),
                                "Duplicate languages detected: %s %s", lang1, lang2
                            )
                            dao.delete(
                                Language::class.java,
                                LanguageTable.FIELD_ID.`in`(*constants(lang1.id, lang2.id))
                            )
                            lang1
                        }
                        .toMutableMap()
                    storeLanguages(json.data, existing)
                }

                dao.updateLastSyncTime(SYNC_TIME_LANGUAGES)
            } ?: return@withContext false
        }
        return@withContext true
    }
}
